barretenberg/cpp/src/barretenberg/vm/avm_trace/AvmMini_execution.cpp

```diff
@@ -0,0 +1,227 @@
+
+        if (Bytecode::has_in_tag(opcode)) {
+            if (pos + AVM_IN_TAG_BYTE_LENGTH > length) {
+                throw std::runtime_error("Instruction tag missing at position " + std::to_string(pos));
+            }
+            in_tag_u8 = bytecode.at(pos);
+            if (in_tag_u8 == static_cast<uint8_t>(AvmMemoryTag::U0) || in_tag_u8 > MAX_MEM_TAG) {
+                throw std::runtime_error("Instruction tag is invalid at position " + std::to_string(pos) +
+                                         " value: " + std::to_string(in_tag_u8));
+            }
+            pos += AVM_IN_TAG_BYTE_LENGTH;
+        }
+
+        auto const in_tag = static_cast<AvmMemoryTag>(in_tag_u8);
+        std::vector<uint32_t> operands{};
+        size_t num_of_operands{};
+        size_t operands_size{};
+
+        if (opcode == OpCode::SET) {
+            switch (in_tag) {
+            case AvmMemoryTag::U8:
+                num_of_operands = 2;
+                operands_size = 5;
+                break;
+            case AvmMemoryTag::U16:
+                num_of_operands = 2;
+                operands_size = 6;
+                break;
+            case AvmMemoryTag::U32:
+                num_of_operands = 2;
+                operands_size = 8;
+                break;
+            case AvmMemoryTag::U64:
+                num_of_operands = 3;
+                operands_size = 12;
+                break;
+            case AvmMemoryTag::U128:
+                num_of_operands = 5;
+                operands_size = 20;
+                break;
+            default:
+                throw std::runtime_error("Instruction tag for SET opcode is invalid at position " +
+                                         std::to_string(pos) + " value: " + std::to_string(in_tag_u8));
+                break;
+            }
+        } else {
+            num_of_operands = Bytecode::OPERANDS_NUM.at(opcode);
+            operands_size = AVM_OPERAND_BYTE_LENGTH * num_of_operands;
+        }
+
+        if (pos + operands_size > length) {
+            throw std::runtime_error("Operand is missing at position " + std::to_string(pos));
+        }
+
+        if (opcode == OpCode::SET && in_tag == AvmMemoryTag::U8) {
+            operands.push_back(static_cast<uint32_t>(bytecode.at(pos)));
+            uint8_t const* ptr = &bytecode.at(pos + 1);
+            uint32_t operand{};
+            serialize::read(ptr, operand);
+            operands.push_back(operand);
+            pos += operands_size;
+        } else if (opcode == OpCode::SET && in_tag == AvmMemoryTag::U16) {
+            uint8_t const* ptr = &bytecode.at(pos);
+            uint16_t operand{};
+            serialize::read(ptr, operand);
+            operands.push_back(static_cast<uint32_t>(operand));
+            ptr = &bytecode.at(pos + 2);
+            uint32_t operand2{};
+            serialize::read(ptr, operand2);
+            operands.push_back(operand2);
+            pos += operands_size;
```

Comment:
Feels like there should be a more generic way to do this :thinking: what about u64 and u128? Do those need else-ifs here?

Reply:
second; are the other cases not included?

Reply:
The critics are absolutely right. Seems convoluted code.

I rewrote and realized that the logic is simpler. We first parse separately operands which are less than 32 bits. This happens for SET and U8/U16.
Then we parse the 32-bit operands under the same umbrella.
I think the new code is much more readable and have more comments.